### PR TITLE
misc: fix "if err != nil { return nil }"

### PR DESCRIPTION
**Pull request by mberhault**
I encountered a random one somewhere and poked around for more.

I **think** this is right. There are no comments in any of them about
not returning an error on errors.

### CODE REVIEW

File: pkg/sql/stats/histogram.go [hunk 1]
```diff
@@ -79,7 +79,7 @@ func EquiDepthHistogram(
 		}
 		encoded, err := sqlbase.EncodeTableKey(nil, upper, encoding.Ascending)
 		if err != nil {
-			return HistogramData{}, nil
+			return HistogramData{}, err
```
Comment (mberhault) [inline pkg/sql/stats/histogram.go]: I assume this is right, no reason not to return an error here.

**Comment by petermattis** [comment]
Didn't we have a real bug that had this pattern before. We should file an issue about adding a lint to catch this.

File: pkg/util/json/encode.go [hunk 2]
```diff
@@ -96,7 +96,7 @@ func (j jsonArray) encode(appendTo []byte) (jEntry uint32, b []byte, err error)
 		var nextJEntry uint32
 		nextJEntry, appendTo, err = j[i].encode(appendTo)
 		if err != nil {
-			return 0, appendTo, nil
+			return 0, appendTo, err
```
Comment (mberhault) [inline pkg/util/json/encode.go]: ditto, but this is an encoding thing, for all I know this may be correct.

File: pkg/sql/sqlbase/testutils.go [hunk 3]
```diff
@@ -106,7 +106,7 @@ func RandDatum(rng *rand.Rand, typ ColumnType, nullOk bool) tree.Datum {
 	case ColumnType_JSON:
 		j, err := json.Random(20, rng)
 		if err != nil {
-			return nil
+			return err
```
Comment (lego) [inline pkg/sql/sqlbase/testutils.go]: Nope, this one is absorbing an error.

File: pkg/util/json/json.go [hunk 4]
```diff
@@ -586,7 +586,7 @@ func (j jsonObject) FetchValKeyOrIdx(key string) JSON {
 func (j jsonArray) FetchValKeyOrIdx(key string) JSON {
 	idx, err := strconv.Atoi(key)
 	if err != nil {
-		return nil
+		return err
```
Comment (lego) [inline pkg/util/json/json.go]: Same, it's absorbing an error.

**Review by a-robinson — COMMENTED** [review]
cc https://github.com/cockroachdb/cockroach/issues/19316

**Review by lgo — COMMENTED** [review]
A few cases there involving Datums. At the moment, the Datum interface doesn't allow returning errors. Only recent additions of JSON now generate errors in here where we never previously had any errors returned. (This is because of a very different behaviour where we lazily decode JSON.) I'm pretty sure @justinj has plans for iterating on this that addresses these cases correctly.

**Comment by tschottdorf** [comment]
Yikes, that's lots of those! Writing the linter isn't technically difficult (we have enough AST traversal precedents in our existing linters), it's just a little tedious and nobody has stepped up to the plate so far.

**Comment by mberhault** [comment]
All in. I had to remove a few that didn't actually have errors returned, so they're just ignored (`dockercluster` and `json`)

**Comment by mberhault** [comment]
Can I get an review?

**Comment by mberhault** [comment]
Ping?

File: pkg/util/json/encode.go (hunk 2 above)
Comment (justinj) [reply pkg/util/json/encode.go]: Yes this is a correct change, thank you!

**Comment by justinj** [comment]
`json` and `sql` changes LGTM

**Review by a-robinson — APPROVED** [review]
`storage/engine/rocksdb.go` LGTM

**Comment by RaduBerinde** [comment]
Wow.. nice finds.

**Comment by knz** [comment]
Reviewed 9 of 12 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/testutils.go, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20440#-L-X-40-6ec8AHhvQu6D-r1-109:-L-YMVhyklVsWDHKyFRO:b-atbqiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f89893f10b3c3cf0ea084be3d3042135d5ad950/pkg/sql/sqlbase/testutils.go#L109)):*
<details><summary><i>Previously, lego (Joey Pereira) wrote…</i></summary><blockquote>

Nope, this one is intentionally absorbing an error.
</blockquote></details>

I would have added a comment to explain why though.

---

*[pkg/util/json/json.go, line 589 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20440#-L-X-40-6ec8AHhvQu6I-r1-589:-L-YMXmd2xWLWUcnXeS_:b6ts7k7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f89893f10b3c3cf0ea084be3d3042135d5ad950/pkg/util/json/json.go#L589)):*
<details><summary><i>Previously, lego (Joey Pereira) wrote…</i></summary><blockquote>

Same, it's absorbing an error.
</blockquote></details>

ditto

@justinj do you mind making a follow-up patch to add a comment here and above that reminds the reader why this is important? So that the next time someone checks there err returns, they don't trip here again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20440)*

**Comment by RaduBerinde** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/testutils.go, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20440#-L-X-40-6ec8AHhvQu6D-r1-109:-L-YSW3y8NDpsfeNIFnk:bkidc24) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f89893f10b3c3cf0ea084be3d3042135d5ad950/pkg/sql/sqlbase/testutils.go#L109)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would have added a comment to explain why though.
</blockquote></details>

This function isn't meant to ever return `nil`. How is `return err` compiling anyway, `err` is not a ` tree.Datum`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20440)*

**Comment by mberhault** [comment]
*[pkg/sql/sqlbase/testutils.go, line 109 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20440#-L-X-40-6ec8AHhvQu6D-r1-109:-L-YUQgIqJYibzDdVIS-:bdkh6w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f89893f10b3c3cf0ea084be3d3042135d5ad950/pkg/sql/sqlbase/testutils.go#L109)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This function isn't meant to ever return `nil`. How is `return err` compiling anyway, `err` is not a ` tree.Datum`?
</blockquote></details>

It's not, I reverted the methods that just returned `JSON`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20440)*

**Pull request merged**